### PR TITLE
KOGITO-4125 - Mongodb persistence generates proto files during codegen

### PR DESCRIPTION
**Pull request by r00ta**
Jira: https://issues.redhat.com/browse/KOGITO-4125

I was surfing around the codebase and I found out that the data index needs the protobuf files, but they are generated only if a persistence module is added as dependency and if `kogito.persistence.type==infinispan`. By consequence, if the user selects `kogito.persistence.type=mongodb` no proto files are generated/exposed and the data-index is not going to work. 

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

**Review by xieshenzh — APPROVED** [review]
@r00ta It looks good to me. 
@swati-kale please also review the change for mongodb persistence codegen. 
Thanks.

**Comment by swati-kale** [comment]
> @r00ta It looks good to me.
> @swati-kale please also review the change for mongodb persistence codegen.
> Thanks.

It looks good to me too.

File: kogito-codegen/src/test/java/org/kie/kogito/codegen/data/AnswerWithAnnotations.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
 import org.infinispan.protostream.annotations.ProtoEnumValue;
 
-public enum AnswerWitAnnotations {
+public enum AnswerWithAnnotations {
```
Comment (r00ta) [inline kogito-codegen/src/test/java/org/kie/kogito/codegen/data/AnswerWithAnnotations.java:20]: Just fixing the typo

**Comment by r00ta** [comment]
Hi @swati-kale , I had to change a bit the code so to make the CI green. Let me know if the changes are still good for you

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/process/persistence/PersistenceGenerator.java [hunk 2]
```diff
@@ -272,8 +271,12 @@ public void updateConfig(ApplicationConfigGenerator cfg) {
         return generatedFiles;
     }
 
+    @SuppressWarnings({"rawtypes", "unchecked"})
     private Collection<GeneratedFile> mongodbBasedPersistence() {
-        Collection<GeneratedFile> generatedFiles = new ArrayList<>();
+
+        ProtoDataClassesResult protoDataClassesResult = protoGenerator.extractDataClasses((Collection) modelClasses);
+        Collection<GeneratedFile> generatedFiles = new ArrayList<>(protoDataClassesResult.getGeneratedFiles());
```
Comment (danielezonca) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/process/persistence/PersistenceGenerator.java:278]: What about move this code outside specific impl methods ([move it here](https://github.com/kiegroup/kogito-runtimes/blob/master/kogito-codegen/src/main/java/org/kie/kogito/codegen/process/persistence/PersistenceGenerator.java#L112))?
I think we can assume proto generation is always needed with persistence (I don't remember if it is always used or only if some addon is enabled)

File: kogito-codegen/src/test/java/org/kie/kogito/codegen/process/persistence/MongoDBPersistenceGeneratorTest.java [hunk 3]
```diff
@@ -56,13 +57,16 @@
...
-                Collections.singleton(Person.class),
-                null,
+                Collections.singleton(GeneratedPOJO.class),
+                new ReflectionProtoGenerator(),
                 null,
                 Arrays.asList("com.mongodb.client.MongoClient"),
                 "mongodb");
         Collection<GeneratedFile> generatedFiles = persistenceGenerator.generate();
 
+        List<GeneratedFile> protoFiles = generatedFiles.stream().filter(gf -> gf.relativePath().contains(".proto")).collect(Collectors.toList());
```
Comment (danielezonca) [inline kogito-codegen/src/test/java/org/kie/kogito/codegen/process/persistence/MongoDBPersistenceGeneratorTest.java:67]: If [this PR](https://github.com/kiegroup/kogito-runtimes/pull/958) will be merged before this one I think it should be nice to create a `PROTO_TYPE` with `RESOURCE` category and use it to filter based on that type. Wdyt?

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/process/persistence/PersistenceGenerator.java (hunk 2 above)
Comment (cristianonicolai) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/process/persistence/PersistenceGenerator.java:278]: actually, the proto files for data index might be needed even if persistence is not used at all. For instance if processes are running in memory but user still wants to track it via events and data index, that should still be possible. So protos should be generated *always* ?
Or perhpas set a Kogito property to allow to disable it if not needed.

Comment (r00ta) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/process/persistence/PersistenceGenerator.java:278]: Hi @cristianonicolai , yes, I agree. If I'm not wrong the operator should not be impacted by this change, but could you confirm @ricardozanini? 
If you agree I would do the change directly in this PR

Comment (ricardozanini) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/process/persistence/PersistenceGenerator.java:278]: Let's always generate the proto files. The operator won't be impacted whatsoever.

File: kogito-codegen/src/test/java/org/kie/kogito/codegen/process/persistence/MongoDBPersistenceGeneratorTest.java (hunk 3 above)
Comment (danielezonca) [reply kogito-codegen/src/test/java/org/kie/kogito/codegen/process/persistence/MongoDBPersistenceGeneratorTest.java:67]: @r00ta 
That PR has been merged so now it is possible to apply the change I proposed. Here is the PR to this branch (with master rebase too)
https://github.com/r00ta/kogito-runtimes/pull/1

Comment (r00ta) [reply kogito-codegen/src/test/java/org/kie/kogito/codegen/process/persistence/MongoDBPersistenceGeneratorTest.java]: Thanks @danielezonca , merged

**Comment by r00ta** [comment]
Waiting for https://github.com/kiegroup/kogito-runtimes/pull/985 before making the changes

**Comment by danielezonca** [comment]
@r00ta 
Linked PR has been merged

**Comment by r00ta** [comment]
Closing this PR since the changes are already included in https://github.com/kiegroup/kogito-runtimes/pull/825

**Pull request closed without merging**